File: panda_joint_reader_node.cpp
//
// Created by sherif on 15.05.22.
//

#include "ros/ros.h"
#include "sensor_msgs/JointState.h"
#include "math.h"
#include <Eigen/Dense>
#include <iostream>
using namespace Eigen;
using namespace std;

void transformationMatrix (double* joint) {
    MatrixXf A1(4,4);
    MatrixXf A2(4,4);
    MatrixXf A3(4,4);
    MatrixXf A4(4,4);
    MatrixXf A5(4,4);
    MatrixXf A6(4,4);
    MatrixXf A7(4,4);
    MatrixXf A8(4,4);


    for (int i=0; i<7 ; i++ ) {
        switch (i) {
            case '0':
                A1 << cos(joint[i]),-sin (joint[i]),0, 0,
                     sin(joint[i]), cos (joint[i]), 0, 0,
                     0, 0, 1, 0.333,
                     0, 0, 0, 1;
                break;

            case '1':
                A2 << cos(joint[i]),0, -sin (joint[i]), 0,
                      sin(joint[i]), 0, cos (joint[i]),0,
                      0, -1, 0, 0,
                      0, 0, 0, 1;
                break;

            case '2':
                A3 << cos(joint[i]),0, sin (joint[i]), 0,
                      sin(joint[i]), 0, -cos (joint[i]), 0,
                      0, 1, 0, 0.316,
                      0, 0, 0, 1;
                break;

            case '3':
                A4  << cos(joint[i]),0, sin (joint[i]), 0.0825*cos(joint[i]),
                        sin(joint[i]), 0, -cos (joint[i]), 0.0825*cos(joint[i]),
                        0, 1, 0, 0,
                        0, 0, 0, 1;
                break;

            case '4':
                A5 << cos(joint[i]),0, -sin (joint[i]), -0.0825*cos(joint[i]),
                       sin(joint[i]), 0, cos (joint[i]), -0.0825*cos(joint[i]),
                       0, -1, 0, 0,
                       0, 0, 0, 1;
                break;

            case '5':
                A6 << cos(joint[i]),0, sin (joint[i]), 0,
                       sin(joint[i]), 0, -cos (joint[i]), 0,
                       0, 1, 0, 0,
                       0, 0, 0, 1;
                break;
            case '6':
                A7 << cos(joint[i]),0, sin (joint[i]), 0.088*cos(joint[i]),
                       sin(joint[i]), 0, -cos (joint[i]), 0.088*cos(joint[i]),
                       0, 1, 0, 0,
                       0, 0, 0, 1;
                break;

        }

    }
    MatrixXf T70 (4,4);
    T70 << A1*A2*A3*A4*A5*A6*A7 ;
    ROS_INFO_STREAM(T70);

}

void callBack (const sensor_msgs::JointState& msg)
{
    double position[6];

    std::cout << "Joints: ";
    for (int i=0; i<7 ; ++i ) {
        position[i] = msg.position[i];
        std:: cout << position[i] << " " ;
    }
    transformationMatrix (position);
    std::cout << "/n/n";
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "panda_joint_reader_node");
    ros::NodeHandle nh;

    std::string topic_name;
    int queue_size;
    nh.getParam ("topic_name", topic_name);
    nh.getParam ("queue_size", queue_size);
    ros::Subscriber sub = nh.subscribe("/joint_states", 10, callBack);
    ros::spin();

    return 0;
}

File: panda_joint_reader_node.py
#!/usr/bin/python3

import rospy
import sys
from sensor_msgs.msg import JointState
import math
import numpy
from sympy import *

init_printing(use_unicode=True)


def callback(msg):
    pos = zeros(1, 7)
    for x in range(len(pos)):
        pos[x] = msg.position
        print(pos)
    mtrans(pos)


def mtrans(joint):
    a1 = numpy.matrix([[math.cos(joint[0]), -math.sin(joint[0]), 0, 0],
                       [math.sin(joint[0]), math.cos(joint[0]), 0, 0],
                       [0, 0, 1, 0.333], [0, 0, 0, 1]])
    a2 = numpy.matrix([[math.cos(joint[1]), 0, -math.sin(joint[1]), 0],
                       [math.sin(joint[1]), 0, math.cos(joint[1]), 0],
                       [0, -1, 0, 0],
                       [0, 0, 0, 1]])
    a3 = numpy.matrix([[math.cos(joint[2]), 0, math.sin(joint[2]), 0],
                       [math.sin(joint[2]), 0, -math.cos(joint[2]), 0],
                       [0, 1, 0, 0.316],
                       [0, 0, 0, 1]])
    a4 = numpy.matrix([[math.cos(joint[3]), 0, math.sin(joint[3]), 0.0825 * math.cos(joint[3])],
                       [math.sin(joint[3]), 0, -math.cos(joint[3]), 0.0825 * math.cos(joint[3])],
                       [0, 1, 0, 0],
                       [0, 0, 0, 1]])
    a5 = numpy.matrix([[math.cos(joint[4]), 0, -math.sin(joint[4]), -0.0825 * math.cos(joint[4])],
                       [math.sin(joint[4]), 0, math.cos(joint[4]), -0.0825 * math.cos(joint[4])],
                       [0, -1, 0, 0],
                       [0, 0, 0, 1]])
    a6 = numpy.matrix([[math.cos(joint[5]), 0, math.sin(joint[5]), 0],
                       [math.sin(joint[5]), 0, -math.cos(joint[5]), 0],
                       [0, 1, 0, 0],
                       [0, 0, 0, 1]])
    a7 = numpy.matrix([[math.cos(joint[6]), 0, math.sin(joint[6]), 0.088 * math.cos(joint[6])],
                       [math.sin(joint[6]), 0, -math.cos(joint[6]), 0.088 * math.cos(joint[6])],
                       [0, 1, 0, 0],
                       [0, 0, 0, 1]])
    finalt = numpy.matmul(a1, a2, a3, a4, a5, a6, a7)
    print(finalt)


def main(argv):
    # Initialize ROS connection
    # The script with the main function needs to be executable.
    # You can make the file executable by executing 'chmod +x /path/to/file.py'
    # in the terminal (without the "'")
    rospy.init_node("panda_joint_reader_node", argv)

    # Read parameters specified in the .launch file
    # The "~" in the beginning means that it will use parameters from the node
    # e.g. /node_name/parameter_name can then be
    # read by rospy.get_param("parameter_name", default)
    topic_name = rospy.get_param("~topic_name", "/default_topic_name")
    queue_size = rospy.get_param("~queue_size", 10)

    # Register a callback function (a function that is called every time a new message arrives)
    rospy.Subscriber(name=topic_name,
                     data_class=JointState,
                     callback=callback,
                     queue_size=queue_size)

    rospy.spin()


if __name__ == '__main__':
    main(sys.argv)
